File: jdba-core-test-db/src/main/java/org/jetbrains/jdba/core/BaseTestHelper.java
package org.jetbrains.jdba.core;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.jdba.exceptions.DBProtectionException;
import org.jetbrains.jdba.exceptions.NoTableOrViewException;
import org.jetbrains.jdba.sql.*;
import org.jetbrains.jdba.util.Collects;

import java.util.Arrays;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;

import static java.lang.String.format;
import static org.jetbrains.jdba.core.Layouts.*;



/**
 * @author Leonid Bushuev from JetBrains
 **/
public abstract class BaseTestHelper<F extends DBFacade> implements DBTestHelper {

  @NotNull
  protected final F db;

  @NotNull
  protected final Scriptum scriptum;

  @NotNull
  protected final Set<String> schemasNotToZap =
      new TreeSet<String>(String.CASE_INSENSITIVE_ORDER);



  protected BaseTestHelper(@NotNull final F db, final @NotNull Scriptum scriptum) {
    this.db = db;
    this.scriptum = scriptum;

    schemasNotToZap.add("information_schema");
  }


  @Override
  public void performCommand(@NotNull final Scriptum scriptum, @NotNull final String commandName) {
    final SqlCommand command = scriptum.command(commandName);
    performCommand(command);
  }

  @Override
  public void performCommand(@NotNull final String commandText) {
    final SqlCommand command = new SqlCommand(commandText);
    performCommand(command);
  }

  @Override
  public void performCommand(@NotNull final SqlCommand command) {
    db.inSession(new InSessionNoResult() {
      @Override
      public void run(@NotNull final DBSession session) {
        session.command(command).run();
      }
    });
  }

  @Override
  public void performCommand(@NotNull final DBTransaction transaction,
                             @NotNull final Scriptum scriptum, @NotNull final String commandName) {
    final SqlCommand command = scriptum.command(commandName);
    performCommand(transaction, command);
  }

  @Override
  public void performCommand(@NotNull final DBTransaction transaction,
                             @NotNull final String commandText) {
    final SqlCommand command = new SqlCommand(commandText);
    performCommand(transaction, command);
  }

  @Override
  public void performCommand(@NotNull final DBTransaction transaction,
                             @NotNull final SqlCommand command) {
    transaction.command(command).run();
  }

  @Override
  public void performScript(@NotNull final Scriptum scriptum, @NotNull final String scriptName) {
    final SqlScript script = scriptum.script(scriptName);
    performScript(script);
  }

  @Override
  public void performScript(@NotNull final SqlScript script) {
    db.inSession(new InSessionNoResult() {
      @Override
      public void run(@NotNull final DBSession session) {
        session.script(script).run();
      }
    });
  }

  @Override
  public void performScript(final String... commands) {
    if (commands == null || commands.length == 0) return;

    SqlScriptBuilder b = new SqlScriptBuilder();
    for (String c : commands) b.add(c);

    performScript(b.build());
  }

  @Override
  public void performCommandOrMetaQueryCommands(@NotNull final Scriptum scriptum,
                                                @NotNull final String operationName) {
    if (scriptum.findText(operationName + "Command") != null) {
      performCommand(scriptum, operationName + "Command");
    }
    else if (scriptum.findText(operationName + "MetaQuery") != null) {
      performMetaQueryCommands(scriptum, operationName + "MetaQuery");
    }
    else {
      throw new IllegalArgumentException(format(
          "The scriptum has no operation %sCommand or %sMetaQuery",
          operationName, operationName));
    }
  }


  @Override
  public void performMetaQueryCommands(@NotNull final Scriptum scriptum,
                                       @NotNull final String metaQueryName,
                                       final Object... params) {
    final SqlQuery<List<String>> metaQuery =
        scriptum.query(metaQueryName, listOf(oneOf(String.class)));

    db.inSession(new InSessionNoResult() {
      @Override
      public void run(@NotNull final DBSession session) {

        List<String> commands = session.query(metaQuery).withParams(params).run();

        SqlScriptBuilder sb = new SqlScriptBuilder();
        for (String command : commands)
          if (command != null && command.length() > 0)
            sb.add(command);

        SqlScript script = sb.build();

        session.script(script).run();

      }
    });
  }


  @Override
  public int countTableRows(@NotNull final String tableName) {
    return db.inSession(new InSession<Integer>() {
      @Override
      public Integer run(@NotNull final DBSession session) {

        return countTableRows(session, tableName);

      }
    });
  }

  @Override
  public int countTableRows(@NotNull final DBTransaction transaction, @NotNull final String tableName) {
    final String queryText = "select count(*) from " + tableName;

    try {
      return transaction.query(queryText, singleOf(Integer.class)).run();
    }
    catch (NoTableOrViewException ntv) {
      return Integer.MIN_VALUE;
    }
  }


  @Override
  public void ensureNoTableOrView(final String... names) {
    final String[][] namePacks = Collects.splitArrayPer(names, 4);
    for (String[] namePack : namePacks) {
      Object[] params;
      if (namePack.length == 4) {
        params = namePack;
      }
      else {
        params = new String[4];
        System.arraycopy(namePack, 0, params, 0, namePack.length);
        Arrays.fill(params, namePack.length, 4, "");
      }

      performMetaQueryCommands(scriptum, "EnsureNoTableOrViewMetaQuery", params);
    }
  }


  @Override
  public void zapSchema() {
    final ConnectionInfo connectionInfo = db.getConnectionInfo();
    final String schemaName = connectionInfo.schemaName;
    if (schemaName != null && schemasNotToZap.contains(schemaName))
      throw new DBProtectionException(format("The schema %s must not be zapped", schemaName), "zapSchema");

    zapSchemaInternally(connectionInfo);
  }

  protected void zapSchemaInternally(final ConnectionInfo connectionInfo) {
    performCommandOrMetaQueryCommands(scriptum, "ZapSchema");
  }


}

File: dekaf-test-db/src/core/BaseTestHelper.kt
package org.jetbrains.dekaf.core

import org.jetbrains.dekaf.exceptions.DBProtectionException
import org.jetbrains.dekaf.exceptions.NoTableOrViewException
import org.jetbrains.dekaf.sql.SqlCommand
import org.jetbrains.dekaf.sql.SqlScript
import org.jetbrains.dekaf.sql.SqlScriptBuilder
import org.jetbrains.dekaf.text.Scriptum
import org.jetbrains.dekaf.util.chopBy
import java.lang.String.format
import java.util.*
import kotlin.reflect.KClass


/**
 * @author Leonid Bushuev from JetBrains
 */
abstract class BaseTestHelper : DBTestHelper {

    protected val db: DBFacade

    protected val scriptum: Scriptum

    protected val schemasNotToZap: MutableSet<String> = TreeSet(String.CASE_INSENSITIVE_ORDER)


    protected constructor(db: DBFacade, helperClass: KClass<out BaseTestHelper>)
        : this(db, Scriptum.of(helperClass))

    protected constructor(db: DBFacade, scriptum: Scriptum) {
        this.db = db
        this.scriptum = scriptum

        schemasNotToZap.add("information_schema")
    }


    override fun fromSingleRowTable(): String {
        return ""
    }

    override fun performCommand(scriptum: Scriptum, commandName: String) {
        val command = scriptum.command(commandName)
        performCommand(command)
    }

    override fun performCommand(commandText: String) {
        val command = SqlCommand(commandText)
        performCommand(command)
    }

    override fun performCommand(command: SqlCommand) {
        db.inSessionDo { session -> session.command(command).run() }
    }

    override fun performCommand(transaction: DBTransaction,
                                scriptum: Scriptum, commandName: String) {
        val command = scriptum.command(commandName)
        performCommand(transaction, command)
    }

    override fun performCommand(transaction: DBTransaction,
                                commandText: String) {
        val command = SqlCommand(commandText)
        performCommand(transaction, command)
    }

    override fun performCommand(transaction: DBTransaction,
                                command: SqlCommand) {
        transaction.command(command).run()
    }

    override fun performScript(scriptum: Scriptum, scriptName: String) {
        val script = scriptum.script(scriptName)
        performScript(script)
    }

    override fun performScript(script: SqlScript) {
        db.inSessionDo { session -> session.script(script).run() }
    }

    override fun performScript(vararg commands: String) {
        if (commands.isEmpty()) return

        val b = SqlScriptBuilder()
        for (c in commands) b.add(c)

        performScript(b.build())
    }

    override fun performCommandOrMetaQueryCommands(scriptum: Scriptum,
                                                   operationName: String) {
        if (scriptum.findText(operationName + "Command") != null) {
            performCommand(scriptum, operationName + "Command")
        }
        else if (scriptum.findText(operationName + "MetaQuery") != null) {
            performMetaQueryCommands(scriptum, operationName + "MetaQuery")
        }
        else {
            throw IllegalArgumentException(format(
                    "The scriptum has no operation %sCommand or %sMetaQuery",
                    operationName, operationName))
        }
    }


    override fun performMetaQueryCommands(scriptum: Scriptum,
                                          metaQueryName: String,
                                          vararg params: Any?) {
        val metaQuery = scriptum.query(metaQueryName, layoutListOf(rowValueOf(String::class.java)))

        db.inSessionDo { session ->
            val commands = session.query(metaQuery).withParams(*params).run()
            if (commands != null && commands.isNotEmpty()) {
                val sb = SqlScriptBuilder()
                for (command in commands) if (command.isNotEmpty()) sb.add(command)
                val script = sb.build()
                session.script(script).run()
            }
        }
    }


    override fun countTableRows(tableName: String): Int {
        return db.inSession { session -> countTableRows(session, tableName) }
    }

    override fun countTableRows(transaction: DBTransaction, tableName: String): Int {
        val queryText = "select count(*) from " + tableName

        try {
            return transaction.query(queryText, layoutSingleValueOf(Int::class.java)).run()!!
        }
        catch (ntv: NoTableOrViewException) {
            return Integer.MIN_VALUE
        }

    }


    override fun ensureNoTableOrView(vararg names: String) {
        val namePacks = names.chopBy(4)
        for (namePack in namePacks) {
            ensureNoTableOrView4(namePack)
        }
    }

    protected open fun ensureNoTableOrView4(params: Array<out String?>) {
        performMetaQueryCommands(scriptum, "EnsureNoTableOrViewMetaQuery", *params)
    }


    override fun zapSchema() {
        val connectionInfo = db.connectionInfo
        val schemaName = connectionInfo.schemaName
        if (schemaName != null && schemasNotToZap.contains(schemaName))
            throw DBProtectionException(format("The schema %s must not be zapped", schemaName), "zapSchema")

        zapSchemaInternally(connectionInfo)
    }

    protected open fun zapSchemaInternally(connectionInfo: ConnectionInfo) {
        performCommandOrMetaQueryCommands(scriptum, "ZapSchema")
    }


}
